src/human_encoding/mod.rs

```diff
@@ -52,4 +54,96 @@ impl<J: Jet> Program<J> {
+            let mut ret = String::new();
+            let expr_width = lines.iter().map(|line| line.expr_str.len()).max().unwrap();
+            let arrow_width = lines.iter().map(|line| line.arrow_str.len()).max().unwrap();
+            let last_line = lines.len();
+            for (n, line) in lines.iter().enumerate() {
+                if skip_before_end && n == last_line - 1 {
+                    ret += "\n";
+                }
+                ret += &format!(
+                    "{0:1$} {2:3$} {4}\n",
+                    line.expr_str, expr_width, line.arrow_str, arrow_width, line.cmr_str
+                );
+            }
+            ret
+        }
+
+        let mut witness_lines = vec![];
+        let mut const_lines = vec![];
+        let mut program_lines = vec![];
+        // Pass 1: compute string data for every node
+        for data in self.root.as_ref().post_order_iter::<MaxSharing<_>>() {
+            let node = data.node;
+            let name = node.name();
+            let mut expr_str = match node.inner() {
+                node::Inner::Fail(entropy) => format!("{} := fail {}", name, entropy),
+                node::Inner::Jet(ref j) => format!("{} := jet_{}", name, j),
+                node::Inner::Word(ref v) => {
+                    format!("{} := const {}", name, serialize::DisplayWord(v))
+                }
+                inner => format!("{} := {}", name, inner),
+            };
+            if let Some(child) = node.left_child() {
+                expr_str.push(' ');
+                expr_str.push_str(child.name());
+            }
+            if let Some(child) = node.right_child() {
+                expr_str.push(' ');
+                expr_str.push_str(child.name());
+            }
+
+            let arrow = node.arrow();
+            let arrow_str = format!(":: {} -> {}", arrow.source, arrow.target).replace('×', "*"); // for human-readable encoding stick with ASCII
+
+            // All witnesses have the same CMR so don't bother printing it
+            let cmr_str = if let node::Inner::Witness(..) = node.inner() {
+                String::new()
+            } else {
+                format!("-- cmr {:.8}...", node.cmr())
+            };
+
+            let print = Print {
+                expr_str,
+                arrow_str,
+                cmr_str,
+            };
+            if let node::Inner::Witness(..) = node.inner() {
+                witness_lines.push(print);
+            } else if let node::Inner::Word(..) = node.inner() {
+                const_lines.push(print);
+            } else {
+                program_lines.push(print);
+            }
+        }
+
+        // Pass 2: actually print everything
+        let mut ret = String::new();
+        if !witness_lines.is_empty() {
+            ret += "-- Witnesses\n";
+            ret += &print_lines(&witness_lines, false);
+            ret += "\n";
+        }
+        if !const_lines.is_empty() {
+            // FIXME detect scribes
```

Comment:
0fe9d92823b1aaddd39ca0ae83374db1884205f0: [I have some code that detects scribe.](https://github.com/uncomputable/hal-simplicity/blob/master/src/compress.rs) I can rebase it and make a PR later if you want.

Reply:
That'd be awesome, though I think we need to decide what to do about const word jets and whether a "scribe" includes them.

Reply:
I'd also hold off on this until the parser is PR'd.